### PR TITLE
Add proxy support for L2

### PR DESCRIPTION
**Pull request by swapnilraj**

### CODE REVIEW

**Comment by hsouf** [comment]
We will have to update the rest of the tests to support proxy

File: test/proxy.test.ts [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    const { decimals } = await owner.call(proxiedTokenContract, 'decimals');
+    expect(decimals).to.equal(4n);
+  });
+
+  it('diasllows random user to upgrade',async () => {
+    try {
+      randomUser.invoke(proxyTokenContract, 'upgrade_implementation',
+          { new_implementation: BigInt(tokenContractB.address) }
+      );
+      expect.fail("Random user cannot upgrade");
```
Comment (hsouf) [inline test/proxy.test.ts:62]: This line keeps on failing tests ` AssertionError: expected 'Random user cannot upgrade' to include 'Proxy: caller is not admin'`. Handling the exception in the catch block should be enough.

Comment (swapnilraj) [reply test/proxy.test.ts:62]: I think its for better logging, but you're right its redundant

Comment (swapnilraj) [reply test/proxy.test.ts]: I think its for logging purposes, you'r right about it being redundant

File: test/proxy.test.ts [hunk 2]
```diff
@@ -56,10 +56,9 @@ describe('Proxy', function () {
 
   it('diasllows random user to upgrade',async () => {
```
Comment (hsouf) [inline test/proxy.test.ts:57]: small typo

**Comment by hsouf** [comment]
Great job! I think this PR is ready to go

**Review by hsouf — APPROVED** [review]

File: test/proxy.test.ts (hunk 2 above)
Comment (swapnilraj) [reply test/proxy.test.ts]: nice catch!

File: test/ETHStaticAToken.test.ts [hunk 3]
```diff
@@ -205,14 +240,15 @@ describe("ETHStaticAToken", function () {
   });
 
   it("claims rewards and mints correct amount of rewards tokens to self", async () => {
-    const user1ClaimableRewards = await l2token.call(
+    const user1ClaimableRewards = await user1.call(
```
Comment (achab) [inline test/ETHStaticAToken.test.ts:243]: why not `await proxiedL2Token.call(...)` ?

File: test/ETHStaticAToken.test.ts [hunk 4]
```diff
@@ -26,85 +31,113 @@ describe("ETHStaticAToken", function () {
...
-      owner: BigInt(bridgeContract.address),
+      owner: BigInt(proxyTokenBridgeL2.address),
     });
+
     //set rewAave address on l2 token bridge
-    await owner.invoke(bridgeContract, "set_reward_token", {
+    await owner.invoke(proxiedTokenBridgeL2, "set_reward_token", {
       reward_token: BigInt(rewAaveTokenL2.address),
     });
+    console.log(l2token.address);
```
Comment (achab) [inline test/ETHStaticAToken.test.ts:90]: remove `console.log(...)` on the line and the one below

File: contracts/src/starknet/rewaave/tokens/IETHstaticAToken.cairo [hunk 5]
```diff
@@ -48,4 +48,14 @@ namespace IETHstaticAToken:
 
     func burn(account : felt, amount : Uint256):
     end
+
+    func initialize_ETHstaticAToken(
```
Comment (achab) [inline contracts/src/starknet/rewaave/tokens/IETHstaticAToken.cairo:52]: why not simply `initialize`?

File: contracts/src/starknet/rewaave/tokens/ETHstaticAToken.cairo [hunk 6]
```diff
@@ -128,6 +110,34 @@ end
 # Externals
 #
 
+@external
+func initialize_ETHstaticAToken{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(
```
Comment (achab) [inline contracts/src/starknet/rewaave/tokens/ETHstaticAToken.cairo:114]: why not simply `initialize`?

File: contracts/src/starknet/rewaave/tokens/ETHstaticAToken.cairo [hunk 7]
```diff
@@ -128,6 +110,34 @@ end
...
+    with_attr error_message("ETHstaticAToken already initialized"):
+        assert name_ = 0
+        assert symbol_ = 0
+        assert decimals_ = 0
+    end
+
+    ERC20_initializer(name, symbol, decimals)
+    ERC20_mint(recipient, initial_supply)
+    Ownable_initializer(controller)
+    # TODO we either need to configure the last_update here, or pause the contract
```
Comment (achab) [inline contracts/src/starknet/rewaave/tokens/ETHstaticAToken.cairo:135]: which of the option below has been selected? is it to be implemented now or in a separate PR?

File: contracts/src/starknet/rewaave/token_bridge.cairo [hunk 8]
```diff
@@ -86,6 +75,20 @@ end
 
 # Externals.
 
+# To finish the init you have to initialize the L2 token contract and the L1 bridge contract.
+@external
+func initialize_token_bridge{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(
```
Comment (achab) [inline contracts/src/starknet/rewaave/token_bridge.cairo:80]: same remark: why not calling this function `initialize`?

File: contracts/src/starknet/rewaave/tokens/ETHstaticAToken.cairo (hunk 6 above)
Comment (swapnilraj) [reply contracts/src/starknet/rewaave/tokens/ETHstaticAToken.cairo:114]: I added the suffix to all the initialize functions otherwise their selector would clash with the one from the proxy

File: test/ETHStaticAToken.test.ts (hunk 4 above)
Comment (swapnilraj) [reply test/ETHStaticAToken.test.ts:90]: ack

File: contracts/src/starknet/rewaave/tokens/ETHstaticAToken.cairo (hunk 7 above)
Comment (swapnilraj) [reply contracts/src/starknet/rewaave/tokens/ETHstaticAToken.cairo:135]: It seems the comment is redundant

**Comment by JorikSchellekens** [comment]
If the tests pass I'm happy

**Review by JorikSchellekens — APPROVED** [review]

File: test/ETHStaticAToken.test.ts [hunk 9]
```diff
@@ -129,15 +160,16 @@ describe("ETHStaticAToken", function () {
...
+    await owner.invoke(proxiedL2Token, "push_acc_rewards_per_token", {
       block: 1,
       acc_rewards_per_token: {
         high: 0,
         low: BigInt(decimalToWad(2)),
       },
     });
 
-    const { acc_rewards_per_token } = await l2token.call(
+    const { acc_rewards_per_token } = await user1.call(
```
Comment (achab) [inline test/ETHStaticAToken.test.ts:171]: `await proxiedL2Token.call("get_acc_rewards_per_token")` should be enough

File: test/tokenbridge.test.ts [hunk 10]
```diff
@@ -302,17 +366,21 @@ describe('TokenBridge', async function() {
...
+    expect(await proxiedL2TokenDai.call('get_last_update', {})).to.deep.equal({ block_number: BigInt(blockNumberDai) });
+    expect(await proxiedL2TokenUsdc.call('get_last_update', {})).to.deep.equal({ block_number: BigInt(blockNumberUsdc) });
+
+    // check balance of L2 tokens
+    expect(await l2user.call(proxiedL2TokenDai, 'balanceOf', { account: BigInt(l2user.starknetContract.address) })).to.deep.equal({ balance: { high: 0n, low:  30n } });
+    expect(await l2user.call(proxiedL2TokenUsdc, 'balanceOf', { account: BigInt(l2user.starknetContract.address) })).to.deep.equal({ balance: { high: 0n, low:  40n } });
   })
 
-  it('L2 users send back reward accrued to L1 user', async () => {
+  it('L2 users sends back reward accrued to L1 user', async () => {
```
Comment (achab) [inline test/tokenbridge.test.ts:377]: you can also change to `L2 user` :)

**Comment by achab** [comment]
very good work, thanks a lot for that! you can still replace `await user.call(contract, params)` with `await contract.call(params)` in the tests. when the CI is fixed, you can safely merge the PR

**Review by achab — APPROVED** [review]
